### ISSUE TITLE
Support user-defined type guard functions

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#1007]
We currently only support some baked-in forms of type checking for type guards -- `typeof` and `instanceof`. In many cases, users will have their own functions that can provide run-time type information.

Proposal: a new syntax, valid only in return type annotations, of the form `x is T` where `x` is a declared parameter in the signature, or `this`, and `T` is any type. This type is actually considered as `boolean`, but "lights up" type guards. Examples:

```ts
function isCat(a: Animal): a is Cat {
  return a.name === 'kitty';
}

var x: Animal;
if(isCat(x)) {
  x.meow(); // OK, x is Cat in this block
}

class Node {
  isLeafNode(): this is LeafNode { throw new Error('abstract'); }
}
class ParentNode extends Node {
  isLeafNode(): this is LeafNode { return false; }
}
class LeafNode extends Node {
  isLeafNode(): this is LeafNode { return true; }
}
var someNode: LeafNode|ParentNode;
if(someNode.isLeafNode()) {
  // someNode: LeafNode in this block
}
```

The forms `if(userCheck([other args,] expr [, other args])) {` and `if(expr.userCheck([any args]))` would apply the type guard to `expr` the same way that `expr instanceof t` and `typeof expr === 'literal'` do today.

**Comment by awerlang** [comment]
@Arnavion I believe that for Array.isArray it would be more correct to express in a non-generic way:

```
interface ArrayConstructor {
    isArray(arg: any): arg is any[];
}
```

This would support union types like `MyCustomObject | any[]`

And how about:

```
interface Object {
    hasOwnProperty<T | any>(v: string): v is T;
}

Object.hasOwnProperty<MyCustomObject>.call(obj, 'id');
```

Or it could narrow to possible matching types `obj` could be:

```
var obj: { id: number; description: string; } | any;
if ('id' in obj) {
  // obj.id is valid here  
}
```

**Comment by Arnavion** [comment]
@awerlang

Actually, I suppose `isArray<T>(): T[]` would give the false assurance that the array only has elements of type T. It might be better to require the user to give an explicit assertion `<T[]>` so that they know they have to perform some validation after the call to isArray to be sure they actually have a T[].

Your last example results in obj being of type any because of union type behavior, so obj.id is also allowed already. If you meant you wanted obj to be considered of the first type in the union (and thus obj.id would have type number) inside the if block, then that doesn't seem right. obj could be `{ id: "foo" }` which would not be assignable to the first type.

For your second example, I assume you meant `hasOwnProperty<T>(obj: T | any, v: string): obj is T;`. This has the same problem - the presence of an 'id' member doesn't really tell you that obj is of MyCustomObject type and not some other type that also has an id property. So it should be something like `hasOwnProperty(obj: any, v: string): obj is { [v]: any };`.

Even if the obj was of type `{ id: number, description: string } | { foo: string }` (i.e., the union doesn't contain any), the second and third example should still result in obj being of type `{ id: any }` inside the if-block, because obj could actually be `{ foo: "5", id: "bar" }` which isn't assignable to the first type but is assignable to the second.

**Comment by awerlang** [comment]
@Arnavion 

I think having `isArray(arg): arg is any[]` is convenient, even though we don't assert element type.

-

About inferring an object type by duck-typing, I consider it is not a problem in its own, although I see the problems it may lead to. This is a concern in #1427 . How would you handle this?

**Comment by Arnavion** [comment]
Sorry, perhaps I wasn't clear. I was agreeing with you that isArray should return `arg is any[]`.

And yes, #1427 is exactly the same as what I said for your other two examples.

**Comment by jbondc** [comment]
This looks very promising as the missing link for #186 

Would also seem like a nice bonus for the concept of an 'interface' alias (see #1945 for patch).
It could simplify the syntax:

```javascript
interface isCat = Animal  where  <(type) => { return type.name === "kitty" }>;
interface isCatOrDog = Animal | Dog where  <(type) => { return type.name === "kitty" || type.name === "dog" }>;

var foo = isCat | string
if (foo instanceof isCat) {

} else {
  // string
}

var foo = isCatOrDog | string
if (foo instanceof isCat) {

} else {
  // string
}
```

emits:

```javascript
if (isCat(foo)) {

} else {
	// string
}

if (isCatOrDog(foo)) {

} else {
	// string
}
```

**Comment by jbondc** [comment]
It's kind of like declaring your own runtime constraints though I don't know if too much magic.

Some symmetry with the 'type alias':
```javascript
type smallString = string  where  <(type) => { return type.length < 10 }>;
type mediumString = string  where  <(type) => { return type.length < 255 }>;

var foo = "hello"
if (typeof foo = "smallString") {
 
} else {
}
```
Emits 
function smallString(type) { return type.length < 10}

```javascript
if (smallString(foo)) {

} else {
    // string
}
```
The downside is it's not really JavaScript... typeof foo = "smallString"

**Comment by jbondc** [comment]
Or this looks a bit better:
```javascript
interface isCat = Animal  where  <instanceof(T) => { return T.name === "kitty" }>;
interface isCatOrDog = Animal where  <instanceof(T) => { return T.name === "kitty" || T.name === "dog" }>;
type smallString = string  where  <typeof(T) => { return T.length < 10 }>;
type mediumString = string  where  <typeof(T) => { return T.length < 255 }>;
```

**Comment by mintern** [comment]
Here's another use case based on a filesystem [Entry](https://developer.mozilla.org/en-US/docs/Web/API/Entry) as defined in the not-yet-standard [File System API](https://developer.mozilla.org/en-US/docs/WebGuide/API/File_System):
```typescript
interface Entry {
    isFile: boolean;
    isDirectory: boolean;
    name: string;
    ...
}

interface DirectoryEntry extends Entry { ... }
interface FileEntry extends Entry { ... }
```

When `(<Entry>entry).isFile`, `entry` is a `FileEntry`, and when `entry.isDirectory`, `entry` is a `DirectoryEntry`.

Note that the `*Entry` classes are not easily accessible, so `instanceof` can't be used.

**Comment by RyanCavanaugh** [comment]
Approved; assignment to Anders is tentative (someone else can give this a shot if they're feeling up to the challenge)

**Comment by tinganho** [comment]
@RyanCavanaugh I decided to take shot on this. Just one question. Can the type guard function take multiple arguments?

```typescript
function isCat(a: Animal, b: number): a is Cat {
  return a.name === 'kitty';
}
```

I'm not sure of its's usefulness with multiple arguments.

**Comment by RyanCavanaugh** [comment]
Multiple arguments should be allowed.

I don't understand the example, though. What are the types of the unbound variables `a` and `b` in the bottom two code blocks?

**Comment by tinganho** [comment]
`a` is of type animal and b just a number.

**Comment by mhegazy** [comment]
Thanks @tinganho!

**Comment by mintern** [comment]
:clap: Thank you!

**Comment by tinganho** [comment]
:smiley:

**Comment by DanielRosenwasser** [comment]
:clap: Great job @tinganho!

**Comment by jbondc** [comment]
I really like this feature but could we instead write it as:
```
interface Sortable {
  sort(): void;
}

class BaseCollection {
  isSortable(): boolean { return false is Sortable; }
}

class List extends BaseCollection implements Sortable {
  isSortable(): boolean { return true is Sortable; }
  sort() { ... }
}

class HashSet extends BaseCollection {
  isSortable(): boolean { return false is Sortable; }
}

class LinkedList extends BaseCollection implements Sortable {
  isSortable(): boolean  { return true is Sortable; }
  sort() { ... }
}
```
The implementation would be roughly the same would instead we attach the identity mapping to the function signature.
This would allow to support:
```
getGroup(animal: Animal): string  {
   switch(animal.breed) {
     case "foxhound":
     case "akita":  
          return "dog" is Dog;
     case "bombay":
     case "birman":  
          return "cat" is Cat;
  }
}
if(getGroup(animal) === "dog") {
  animal; // Dog
} else {
   animal; // Cat (we know this because "dog" was removed from "dog" | "cat"
}
```

**Comment by JsonFreeman** [comment]
@jbondc The scenario you are describing is covered by #1003, not this feature.

**Comment by jbondc** [comment]
I think the boolean/predicate is just a special case:
```
if( something(a) ) {
   a;
}
```
Infer the constant on the RHS to be true:
```
if( something(a) === true ) {
   a;
}
```
The general rule for type narrowing by constant value is:
```
let rhsValue = true;
If a constant literal is found is the RHS and the type of the LHS is a 'set type', 
  if the function something has an identity mapping of constant values
       if the function has a mapping constantToType[rhsValue]
            if( isAssignable(constantToType[rhsValue], a)  ) {
                   // then narrow a is constantToType[rhsValue]
             }
```
It's not too concrete right now, but fairly confident the end result is the same without two different narrowing algorithms.

**Comment by JsonFreeman** [comment]
I see what you are saying, but it requires a way for the user to express the intended mapping from return values to narrowing types. I think this is not a useful capability in general. Certain special cases are useful though. Type predicates and property accesses with literals are indeed useful cases. They can use different mechanisms because they are different enough.

**Comment by rotemdan** [comment]
I'm aware the feature has already been finalized in 1.6, but just wanted to mention a possible alternative syntax using decorators (extended to functions [at least for this case] and evaluated at compile-time):

```ts
@inquiresType(Cat)
function isCat(a: Animal): boolean {
  return a.name === 'kitty';
}

var x: Animal;
if(isCat(x)) {
  x.meow(); // OK, x is Cat in this block
} 

class Node {
  @inquiresClassType(LeafNode)
  isLeafNode(): boolean { throw new Error('abstract'); }
}

class ParentNode extends Node {
  @inquiresClassType(LeafNode)
  isLeafNode(): boolean  { return false; }
}

class LeafNode extends Node {
  @inquiresClassType(LeafNode)
  isLeafNode(): boolean { return true; }
}

var someNode: LeafNode|ParentNode;
if(someNode.isLeafNode()) {
  // someNode: LeafNode in this block
}
```
I'm not saying the current syntax is bad at all. However, I would imagine that if the feature was proposed today, an approach like this might be seen as more conservative (would have a lower general impact on syntax, and probably be easier to implement in the compiler).

**Comment by use-strict** [comment]
@RyanCavanaugh , the first example says you can use `this is LeafNode`, but in reality only the `a is Cat` syntax works.

Maybe the example wasn't updated or it's a bug. The merged PR description however leads me to believe this was never implemented.

**Comment by RyanCavanaugh** [comment]
@use-strict you're correct that it wasn't implemented. Probably best to open a new issue if this is something you'd like to have (so we can gauge how many people are running into it).

**Comment by mrcrowl** [comment]
@RyanCavanaugh, did support for `this is X` make it into the 1.6 release?
```typescript
class ParentNode extends Node {
  isLeafNode(): this is LeafNode { return false; }
}
```
I'm doing something along the lines of the following, but it appears to be unsupported?
```typescript
abstract class Animal
{
    get isCat(): this is Dog;
}

class Dog extends Animal
{
    get isDog(): this is Dog { return true; }
}
```

**Comment by eggers** [comment]
Show this have solved the error `Index signature of object type implicitly has an 'any' type.` for the example below?

```
var foo = {bar: 'bar'}
var prop = 'bar';

console.log(foo[prop])
```

**Comment by RyanCavanaugh** [comment]
@eggers You'd have to rewrite it slightly, but it works:

```ts
function hasProp<T>(x: T, name: string): x is T & { [key: string]: any } {
  return x && x.hasOwnProperty(name);
}

var foo = {bar: 'bar'}
var prop = 'bar';

if (hasProp(foo, prop)) {
  console.log(foo[prop])
}
```

**Comment by eggers** [comment]
@RyanCavanaugh Not quite, because then the below doesn't throw an error like it should:

```ts
function hasProp<T>(x: T, name: string): x is T & { [key: string]: any } {
  return x && x.hasOwnProperty(name);
}

var foo = {bar: 'bar'}
var prop = 'bar';

if (hasProp(foo, prop)) {
  console.log(foo['abc'])
}
```

**Comment by shelby3** [comment]
I don't understand why the proposed syntax is necessary, nor how it is even sound typing. The compiler isn't able to check the typing assumption.

@RyanCavanaugh [wrote](https://github.com/Microsoft/TypeScript/issues/1007#issuecomment-61241232):
> Perhaps the second example was not clear in its intent. Consider something like this, where instanceof would not work:

The following works and there is no unchecked typing assumption:

abstract class Sortable {
  abstract sort(): void
}

class BaseCollection {
}

class List extends BaseCollection implements Sortable {
  sort() {}
}

var someNode : BaseCollection | List

if(someNode instanceof Sortable) {
  someNode.sort();
}

**Comment by kitsonk** [comment]
@shelby3 why are you commenting on issues that have been closed for 18 months including picking out comments made nearly two years ago?  If you have an issue, it would be far more productive to state it against the current state of TypeScript instead of spending energy trying to debate things that have been dead a buried for extended periods of time.

**Comment by shelby3** [comment]
@kitsonk wrote:
> why are you commenting on issues that have been closed for 18 months including picking out comments made nearly two years ago?

Because the FAQ [tells me to do precisely that](https://github.com/Microsoft/TypeScript/wiki/FAQ#i-disagree-with-the-outcome-of-this-suggestion):

> Denial: It's healthy to believe that a suggestion might come back later. Do keep leaving feedback! We look at all comments on all issues - closed or otherwise. If you encounter a problem that would have been addressed by a suggestion, leave a comment explaning what you were doing and how you could have had a better experience. Having a record of these use cases helps us reprioritize.

@kitsonk wrote:
> If you have an issue, it would be far more productive ***to state it against the current state of TypeScript*** instead of spending energy trying to debate things that have been dead a buried for extended periods of time.

The bolded is exactly what I did. A bad design decision is never buried. In particular, I am hoping Google SoundScript team is reading, so they will know which mistakes to consider avoiding.

And again, I asked for someone to provide to me justification for this feature change. Until I understand some counter logic to my points, I am going to continue to think it was a bad decision, hope it gets deprecated, and also try to influence it not being duplicated in other derivative works.

I am also open to being convinced via discussion that I am wrong, which I stated.

Btw, are you angry? I will quote from that same FAQ:

> Anger: Don't be angry.

The way you worded your response to me was to entirely ignore the substance of my discussion and basically it seems you are trying to demoralize me by belittling my sincere and thoughtful contribution. A more productive (and less politically contentious) response from you would have actually address the substance of my comments. Any way, I will continue to do what I do, which is operate with a high standard of excellence and ethics and ignore those who try to turn factual discussions into political contests.

**Comment by kitsonk** [comment]
@shelby3 I am trying to point out that your thoughtful comments will simply be lost and ignored.  Even though the FAQ states that, I think the context of which your comments were directed is for an issue that was to introduce user defined type guards.  You seem to want to debate the nature of the nominal typing in TypeScript.  I noticed you commented on #202 which is likely more aligned to the topic at hand, than some issue that has been resolved.

It was not politically motivated at all, nor am I angry.  I saw someone who appears to be relatively new to TypeScript add significant comments to several older issues, some of which didn't seem to be aligned to the topic at hand.  While maybe a bit direct, I was suggestion ways that individual might get more appropriate consideration to their thoughts.

**Comment by shelby3** [comment]
@kitsonk thanks for clarifying that you are not waging any political campaign against me. To clarify the way I operate, I respond and document my own thoughts in context where they apply, for my own record of my thoughts and design decisions. If my sharing benefits anyone else and leads to any mutually beneficial discussion, then great. But driving action directly from my effort to study issues and comment is not the sole objective of the way I am working. For example, I will cross-reference in newer discussions, comments on older issues that I've read and commented on to try to bring myself up to speed on TypeScript. So just because a comment goes here in context, doesn't it mean it won't get tied into discussion impacting new discussion. Yes I am very interested in the nominal typing issue and in particular typeclasses and probably about to make a proposal on add them to model the prototype chain (but this is very complex). I did add an issue today suggesting `if-else` and `switch` as expressions. I am all over the place on TypeScript for the past 36 hours or so, which I just noticed [you are commenting negatively on](https://github.com/Microsoft/TypeScript/issues/10841#issuecomment-246127278).

**Comment by RyanCavanaugh** [comment]
Just to set expectations here, there is absolutely no way we're going to *remove* an oft-requested and popular type system feature from the language because some people feel it can't be used correctly. You can set up a TSLint rule to disallow it in your codebase if you think it's somehow dangerous.

We use type guards internally all over the place, e.g. https://github.com/Microsoft/TypeScript/blob/master/src/compiler/utilities.ts#L894 . There are lots of situations where they're useful and we added this because sometimes the built-in type checking things are not good enough.

**Comment by shelby3** [comment]
@RyanCavanaugh wrote:
> there is absolutely no way we're going to _remove_ an oft-requested and popular type system feature from the language because some people ***feel it can't be used correctly***.

I appreciate your bluntness.

There are no *feelings* involved in [my *objective* analysis of the inability to the use the feature correctly](https://github.com/Microsoft/TypeScript/issues/1007#issuecomment-246085962). I would love to see someone refute my logic.

I explained that:

1. Afaics, the compiler can't check the soundness of the declaration, thus (if my analysis is correct?) it is impossible to use the feature correctly. What is the point of a compile-time check on types, if type declarations can't be checked by the compiler to be consistent?

It is unfathomable to me that you would think that depending on users to manually check the consistency of their declarations is acceptable for compiler-enforced type checking. It just boggles my mind how this feature ever got into the language. I must be missing something and I am hoping someone can point out my myopia? I hope I am wrong, because my current state-of-mind is bewilderment as to how a team can produce such a great project with many wise decisions and let this very bad one slip through the cracks.

2. I can't find any use for the feature that can't be accurately checked by the compiler with an equally eloquent syntax and which doesn't require this compiler uncheckable feature.

It really boggles my mind. I am still hoping someone can point out how I am missing the point of this feature.

**Comment by aluanhaddad** [comment]
@shelby3 
> The following works and there is no unchecked typing assumption:
> 
> abstract class Sortable {
>   abstract sort(): void
> }
> 
> class BaseCollection {
> }
> 
> class List extends BaseCollection implements Sortable {
>   sort() {}
> }
> 
> var someNode : BaseCollection | List
> 
> if(someNode instanceof Sortable) {
>   someNode.sort();
> }

It does not work. Checkout the generated JavaScript and the behavior of the code as demonstrated here http://www.typescriptlang.org/play/#src=abstract%20class%20Sortable%20%7B%0D%0A%20%20abstract%20sort()%3A%20void%0D%0A%7D%0D%0A%0D%0Aclass%20BaseCollection%20%7B%0D%0A%7D%0D%0A%0D%0Aclass%20List%20extends%20BaseCollection%20implements%20Sortable%20%7B%0D%0A%20%20sort()%20%7B%7D%0D%0A%7D%0D%0A%0D%0Avar%20someNode%20%3A%20BaseCollection%20%7C%20List%0D%0A%0D%0Aif(someNode%20instanceof%20Sortable)%20%7B%0D%0A%20%20alert('someNode%20has%20Sortable%20in%20its%20prototype%20chain')%3B%0D%0A%7D%20else%20%7B%0D%0A%09alert%20('someNode%20does%20NOT%20have%20Sortable%20in%20its%20prototype%20chain')%0D%0A%7D

This however, works quite well.
```TypeScript
abstract class sortable {
    abstract sort(): void;
}
class BaseCollection {
}
class List extends BaseCollection implements Sortable {
  sort() {}
}

var someNode : BaseCollection | List

function isSortable(collection: BaseCollection): collection is Sortable {
    return typeof collection.sort === 'function';
}

if(isSortable(someNode)) {
    someNode.sort();
}
```

It is my strong feeling that this example demonstrates the usefulness of user defined type guards.


> There are no feelings involved in my objective analysis of the inability to the use the feature correctly. I would love to see someone refute my logic.

You are begging the question.

**Comment by shelby3** [comment]
@aluanhaddad I was missing the construction of an instance for `someNode`:

```ts
var someNode : BaseCollection | List   // someNode is `undefined`

if(someNode instanceof Sortable) {
     someNode.sort();
}
```
My [corrected example](http://www.typescriptlang.org/play/#src=abstract%20class%20Sortable%20%7B%0D%0A%20%20abstract%20sort()%3A%20void%0D%0A%7D%0D%0A%0D%0Aclass%20BaseCollection%20%7B%0D%0A%7D%0D%0A%0D%0Aclass%20List%20extends%20Sortable%20%7B%0D%0A%20%20sort()%20%7B%7D%0D%0A%7D%0D%0A%0D%0Avar%20someNode%20%3A%20BaseCollection%20%7C%20List%20%3D%20new%20List()%0D%0A%0D%0Aif(someNode%20instanceof%20Sortable)%20%7B%0D%0A%20%20alert('someNode%20has%20Sortable%20in%20its%20prototype%20chain')%3B%0D%0A%7D%20else%20%7B%0D%0A%09alert%20('someNode%20does%20NOT%20have%20Sortable%20in%20its%20prototype%20chain')%0D%0A%7D) shows that 100% `abstract` classes can be put into the `prototype` chain by employing the `extends` keyword.

The remaining problem is that TypeScript does not support multiple inheritance of subclass through [linearization of the hierarchy](https://www.google.com.ph/search?q=linearization+Scala), so we can't currently `extend` from both `BaseCollection` and `Sortable` simultaneously.

It appears to me to be an error that TypeScript is not putting the implemented interfaces in the `prototype` chain so that `instanceof` will work correctly for interfaces. Even though (even `abstract`) classes are treated nominally for the `prototype` chain with `extend`, they are still structurally typed otherwise. So the argument that interfaces are structurally typed doesn't seem to a contradiction.

So I maintain my stance that the feature of this issue was not necessary. And now I add a bug report that TypeScript is failing to put interfaces into `prototype` chain. Not fixing a bug and making an unnecessary unsound typing feature to compensate for not doing so, it a very bad design mistake. That is what open source is for, so that with enough eyeballs someone will catch the error.

**Comment by shelby3** [comment]
@shelby3 wrote:
> Even though (even `abstract`) classes are treated nominally for the `prototype` chain with `extend`, they are still structurally typed otherwise. So the argument that interfaces are structurally typed doesn't seem to be a contradiction.

I suppose the argument against my logic would be that putting interfaces in the `prototype` chain would be [metadata inserted by the compiler and not an erasure of all elided typing hints](https://github.com/Microsoft/TypeScript/issues/10841#issuecomment-246172877).

The feature of this issue (which I argue is unsound) burdens the user will inserting metadata and eliminates the ability of the compiler to check it. Thus on balance, I think my logic stands. The compiler should insert the metadata and thus insure soundness.

**Comment by mintern** [comment]
Here's a perspective from a huge fan of TypeScript who is not involved in its development. We had a huge JavaScript codebase that we converted to TypeScript, and it's now much easier to work with. We now have easier and safer refactoring, auto-completion, semantic navigation, and some compiler checks on annotations that used to be part of comments instead of being part of the code.

@shelby3 wrote:

> Afaics, the compiler can't check the soundness of the declaration, thus (if my analysis is correct?) it is impossible to use the feature correctly. What is the point of a compile-time check on types, if type declarations can't be checked by the compiler to be consistent?

Because of the [incompleteness theorem](https://en.wikipedia.org/wiki/G%C3%B6del%27s_incompleteness_theorems), there is always a correct program that cannot be described by a rigid type system. Every language that I know of has an escape hatch: Rust has `unsafe`, Java has reflection and type-casting, and the list goes on and on. More to the point, raw JavaScript is completely dynamic, and it's up to the programmer to write everything correctly, with no help from a compiler.

To me, TypeScript is a tool that helps me write correct programs. Its gradual typing is a hint about its philosophy: start with whatever you have, and improve it incrementally by adding information to your program. As you continue to do that, TypeScript helps you more and more. I don't think its goal is to be perfect, just to be better, and it has succeeded for us at least: the TypeScript is significantly better than the JavaScript it replaced for us.

Bringing it back to this discussion, user-defined type guards are a perfect example. I can write a little function (like **aluanhaddad**'s `isSortable` example) that tells the compiler, "Hey, I know that you can't tell this function is actually checking for a `Sortable` type, but trust me, I know what I'm doing here." And the compiler will believe me and allow me to use this new `isSortable` tool in the rest of my program.

Without that feature, I have to write the check and cast anywhere in my program that I need to check for a `Sortable` type, and that's much more error-prone than a user-defined type guard.

In other words, you're right that the compiler isn't verifying the correctness of a user-defined type guard function, but I'm pretty sure that's exactly the point of the feature.

**Comment by shelby3** [comment]
@mintern wrote:
> Every language that I know of has an escape hatch:

But my point is that the feature added from this issue is not an escape hatch from the compiler. We adding a feature to the compiler to check something that it would not otherwise be able to check, and we are enabling a human to tell the compiler what the correct type of that check should be without enabling the compiler to have any way to verify that the human didn't make a mistake. Rather just leave the checking off and have a true escape hatch, e.g. the type `any`, if the compiler can't check it.

Also my point is that the compiler could indeed check it properly but isn't doing, so instead we create an unnecessary feature which is more unsound than not having the feature at all and leaving it uncheck (an unverified check is worse than no check, because it is an assurance that be totally deceiving).

**Comment by aluanhaddad** [comment]
@shelby3 
> I realize you are a younger man well before the precipitous drop in testosterone (starting university in 2007 versus mine in 1983) and I notice you are upvoting spiteful comments against me in other issue threads where I am participating, but really man I am 51 years old. Ego is for little people. You'll earn your stripes eventually and realize the wisdom of the linked blog.

These statements have no place in this discussion.

> Btw, I appreciate your discussion (even the humorous unnecessary bravado part). You are helping me (we are helping each other) to understand the solution. Isn't that wonderful.

Indeed, I was attempting to be humorous, but I was also being serious. Stating that one is objective and not coming from any emotional place, especially after having already intimated that others are coming from an emotional place, makes one's further statements highly suspect.
Yes, we are helping each other learn, and yes that is a wonderful thing.


> So I maintain my stance that the feature of this issue was not necessary. And now I add a bug report that TypeScript is failing to put interfaces into prototype chain. Not fixing a bug and making an unnecessary unsound typing feature to compensate for not doing so, is a design mistake. That is what open source is for, so that with enough eyeballs someone will catch the error.


> The remaining problem is that TypeScript does not support multiple inheritance of subclass through linearization of the hierarchy, so we can't currently extend from both BaseCollection and Sortable simultaneously.

In TypeScript interfaces have always been structural. They declare a compile time name for a certain shape. Not all TypeScript/JavaScript programs are class based. Consider the following code.
```TypeScript
export function openModal(modalOptions: ModalOptions): void {
    const modal = createModalImpl(modalOptions.template, modalOptions.viewModel);
    modal.onOpen = modalOptions.onOpen;
    modal.onClose = modalOptions.onClose;
    modal.editable = modalOptions.readonly;
    modal.open();
}

export interface ModalOptions {
    template: string;
    viewModel: any;
    onOpen: (callback: () => void);
    onClose: (callback: () => void);
    readonly: boolean;
}
```
The purpose of the ModalOptions interface is to declare part of the interface of the _function_ `openModal` so that the TypeScript compiler can check that the correct arguments are specified and so that the caller knows what to pass. It is used for both documentation and typechecking to formally describe the contract of the API, no classes are involved. Thus, the following two examples have the same behavior.

```TypeScript
import { openModal, ModalOptions } from 'basic-modal-utilities';
import { createViewModel } from './view-model';
import template from './template.html';

// ex 1.
const vm = {
    name: 'Bob',
    friends: ['Alice', 'Eve']
};
openModal({
    template: template,
    viewModel: vm,
    readonly: true,
    onOpen: {
        console.log(`modal opened with data ${json.stringify(vm).}`);
    },
    onClose: () => {
        if (vm.name !== 'Bob' || 
            vm.friends[0] !== 'Alice' || 
            vm.friends[1] !== 'Eve' || 
            vm.length > 2) {
            console.error('Test failed: viewModel was modified');
        }
    }
});

// ex 2. 

class ReadonlyModalOptions implements ModalOptions {
    readonly = true;

    constructor(private viewModel, private template) { }

    onOption() {
        console.log(`modal opened with data ${json.stringify(this.viewModel)}.`);   
    }

    onClose: () => {
        if (this.viewModel.name !== 'Bob' ||
            this.viewModel.friends[0] !== 'Alice' || 
            this.viewModel.friends[1] !== 'Eve' || 
            this.viewModel.length > 2) {
            console.error('Test failed: viewModel was modified');
        }
    }
}
const vm = {
    name: 'Bob',
    friends: ['Alice', 'Eve']
};
openModal(new ReadonlyModalOptions(vm, template));
```
Lots of APIs work this way. They simply require that a caller pass arguments that meets their requirements, but not how those requirements are defined. This is not only a common pattern but a useful one. If interfaces were added to the prototype chain, then it would be reasonable for someone to add the following line of code to the `openModal` function:
```
if (!modalOptions instanceof ModalOptions) {
    throw TypeError();
}
```
This introduces serious problems. Ex 1. no is longer valid. This is (annoying, but not fatal) for TypeScript consumers who do not want to create a class, but it is catastrophic for JavaScript consumers that can not implement the interface without manually wiring up the prototype chain themselves. 

Another issue with linearization is that, as more interfaces are implemented, the depth of the synthetic class hierarchy grows exponentially. This becomes especially problematic in cases where interfaces extend other interfaces and classes reimplement interfaces. 
Consider
```TypeScript
interface Sortable<T> {
    sort<U>(sortBy: (x: T) => U): this;
}
interface NumericallyIndexable<T> {
    [index: number]: T;
}
interface ArrayLike<T> extends Sortable<T>, NumericallyIndexable<T> {
    map<U>(projection: (x: T) => U): ArrayLike<U>;
    filter(predicate: (x: T) => boolean): ArrayLike<T>;
}
interface Grouping<K, V> {
   get(key: K): ArrayLike<V>;
}
class RichCollection implements ArrayLike<T>, Sortable<T> {
    constructor(private elements: T[] = []);
    sort<U>(sortBy: (x: T) => U) { return this. }
    map<U>(projection: (x: T) => U) { return new RichCollection(this.elements.map(projection)); }
    filter(predicate: (x: T) => boolean) { return new RichCollection(this.elements.filter(predicate)); }
    groupBy<K>(keySelector: (x: T) => K): Grouping<K, T> { ... }
}
```

**Comment by shelby3** [comment]
@aluanhaddad wrote:
> These statements have no place in this discussion.

Neither do these provocateurs:

> It is my strong feeling that this example demonstrates the usefulness of user defined type guards.

>> There are no feelings involved in my objective analysis of the inability to the use the feature correctly. I would love to see someone refute my logic.

> You are begging the question.

When you intentionally start a fight, then don't be surprised when you get punched.

**Comment by aluanhaddad** [comment]
@shelby3
> Listen son, I don't know what your problem is, but before you make a false accusation, please remember to quote the source of your false accusation. I made no such precedent statement about others. I quoted from the FAQ. Direct any such accusation to the source.

In https://github.com/Microsoft/TypeScript/issues/1007#issuecomment-246127188
you wrote
>Btw, are you angry? I will quote from that same FAQ:
 
suggesting that @kitsonk was angry. There was no evidence for that. Even if there was you were asking the question rhetorically, which is weakens your argument by introducing an ad hominem fallacy.

**Comment by spion** [comment]
@shelby3 You are wrong. This sort of feature is precisely what makes TypeScript different from other typed languages. While they decide NOT to trust the user unless the user's code fits their narrow preconceptions of what is correctly modelled code, TypeScript takes a different approach. TypeScript mostly trusts the user. The user, in turn, needs to take special care in ensuring *some* of their code is correct, like type guards. But not all, as it would be with dynamic languages. And this is the biggest win of TypeScript: it lets you pick between "great power, great responsibility" (carefully hand-check the code, assert that it produces the correct types) and "lesser power, lesser responsibility" (once you hand-"prove" type assertions are correct, you don't have to do it for the rest of the code).

Its not a type system that ensures correctness. Its a type system that works with the user to ensure it does.

**Comment by kitsonk** [comment]
> Its not a type system that ensures correctness. Its a type system that works with the user to ensure it does.

I agree, while acknowledging that the underlying language, JavaScript is permissive and loosely typed.  Specifically quoting another TypeScript non-goal:

> 3) Apply a sound or "provably correct" type system. Instead, strike a balance between correctness and productivity.

**Comment by shelby3** [comment]
@aluanhaddad:
>> Btw, are you angry? I will quote from that same FAQ:

Do you not see question mark? Are you just determined to ignore the difference in the English language between a question and an implication or accusation or even insinuation.

> There was no evidence for that.

Perhaps you should remember to read the entire comment:

> @kitsonk wrote:
>> why are you commenting on issues that have been closed for 18 months including picking out comments made nearly two years ago?

...

>> Btw, are you angry? I will quote from that same FAQ:
>
> Anger: Don't be angry.

And the follow-up:

> @kitsonk thanks for clarifying that you are not waging any political campaign against me. To clarify the way I operate, I respond and document...

Why are you wasting precious time and carrying around baggage. Let it go and focus on the technical discussion. I will not respond again on this off-topic noise.

**Comment by aluanhaddad** [comment]
@shelby3 I do not think ageist or sexists remarks have any place in a technical discussion. I felt compelled to object.

Regardless, perhaps you would care to respond to the technical side of my comment, specifically the concerns I raised about linearization of interfaces into the class hierarchy.

**Comment by shelby3** [comment]
@aluanhaddad:
> @shelby3 I do not think ageist or sexists remarks have any place in a technical discussion.

OMG! He pulls out the political correctness legal discrimination and State-monopoly-on-force weapon. I see you want to build a case for a ban or what ever.

I will have no more discussion with you. You just earned a ban from me. You can speak to my attorney.

**Comment by shelby3** [comment]
@spion wrote:
> @shelby3 You are wrong.

Wrong about what? Please quote my statement that is "wrong". Rather I think you mean to say that TypeScript has design goals which don't match my concerns. That doesn't make me wrong. You know very well from other thread where you are participating, that I have stated that TypeScript's goals seem somewhat ambiguous to me.

The false accusations are flying in and now you @spion are downvoting my comments. So I see it has turned into a "shoot the messenger" politics instead of carefully thought out non-acriminous discussion of technology. That means it is digressing and it is about time for me to leave unless the level of maturity and ethics rises pronto.

**Comment by spion** [comment]
No it hasn't. You started to veer off the technical rail by discussing age and testosterone, then responded with "Let it go and focus on the technical discussion. I will not respond again on this off-topic noise". 

You are inconsistent, rude, abrasive. Hell, you even sliced my comment at the "you are wrong" point and proceeded to ignore the rest of it, pretending it doesn't exist. 

I suggest you use TypeScript for at a couple of weeks before commenting further on it.

**Comment by mintern** [comment]
@shelby3 I'm not going to say any more than this: from an outsider perspective, your tone is consistently aggressive. Your comments seem (to me!) to be dripping with anger. I think part of it is words like "son", which many people consider to be demeaning. Meanwhile, I feel like everyone else has been patiently trying to describe TypeScript's motivations, design decisions, and philosophy.

Anger and aggression in a discussion like this will convince no one.

I think you're probably right when you say that TypeScript has design goals that don't match your concerns. Maybe you could fork it and create a new language that linearizes interfaces in the way you describe? That sounds like it could be interesting, although it would be a rather different language from TypeScript, and that's OK.

**Comment by shelby3** [comment]
> You started to veer off the technical rail by discussing age and testosterone

That was a response to equally abusive provocateur verbiage. If you continue this politcal grandstanding, you will go on my block also. Stay on the technical topic, otherwise this discussion ends. (which is probably what you all are trying to accomplish any way)

**Comment by shelby3** [comment]
@mintern wrote:
> Maybe you could fork it and create a new language that linearizes interfaces in the way you describe?

Please re-read my comments more carefully:

**Comment by shelby3** [comment]
@spoin wrote:
> Its not a type system that ensures correctness. Its a type system that works with the user to ensure it.

Then it is not a type system. A type system by definition is:

> In programming languages, a type system is a collection of rules that assign a property called type to various constructs a computer program consists of, such as variables, expressions, functions or modules.[1] The main purpose of a type system is to reduce possibilities for bugs in computer programs[2] by defining interfaces between different parts of a computer program, and ***then checking that the parts have been connected in a consistent way***.

> A type system associates a type with each computed value and, by examining the flow of these values, attempts to ensure or prove that no type errors can occur.

So if you introduce human error in between the declarations and the checking of their consistent interaction, then you have defeated typing.

**Comment by aluanhaddad** [comment]
Most languages have type systems are not 100% sound. They have various constructs which bypass or overrule the type checker. Some examples of this include covariant mutable collections, arbitrary casts (sometimes by way of intermediate casts), and generic erasure, and Scalas's `uncheckedVariance` annotation. 

TypeScript's type system is not and does not claim to be 100% sound, and it has many of the unsound constructs above, arguably for very practical reasons. 

However, structural typing is not an unsoundness it is a deliberate choice which specifies the semantics of what it means to say that a value `x` is of a type `T`.

TypeScript was designed in part to formally specify JavaScript's type system. JavaScript does not have manifest type system. `typeof` is a type level operator but it returns a string because it cannot return a type. People generally regard `instanceof` as a type level operator but in fact both of its operands are always values.
JavaScript is effectively duck typed language. That it is interested in the shape and not the heritage of objects. 

Structural typing allows TypeScript to model this very effectively.

**Comment by shelby3** [comment]
How many times am I going to have to repeat the distinction between bypassing (i.e. turning off) the typechecker, versus introducing unchecked values into types violating consistency of types. Turning off typing is not the same as infiltrating the types with unchecked values that aren't congruent with those types. Once you take values out of the type system, those values can never safely come back into the type system, unless the compiler can perform a runtime verification of the value and throw an exception.

**Comment by shelby3** [comment]
@mintern wrote:
> I agree, while acknowledging that the underlying language, JavaScript is permissive and loosely typed. Specifically quoting another TypeScript non-goal:

>> 3) Apply a sound or "provably correct" type system. Instead, strike a balance between correctness and productivity.

I interpreted that design goal to mean that the type system wouldn't check everything and that absolute runtime soundness isn't a realistic goal for the full range of the JavaScript universe (libraries, frameworks, etc) as it is today. And I agree.

I did not interpret it to mean that we should purposefully inject unenforced values into the type system to destroy the internal consistency of the type system, potentially turning it into random soup of contagion of the interaction of multiple heuristics in an eventual exponential explosion of Complexity of Whoops.

**Comment by shelby3** [comment]
Thinking about if the feature enabled by this issue is not first-class, and thus unable to infiltrate the rest of the type system, thus cordoning the impact of any human error. It appears to not be first-class, except perhaps for example maybe the narrowed type can be used along with partial function application (does TypeScript have partial function application?) to select which of an overloaded function (or method) is saved to a callback instance, which is then passed around as a first-class function which can leak into any part of the type system.

Also thinking about @aluanhaddad's point that when interopting with JavaScript that didn't populate the `prototype` chain with interfaces per my alternative suggestion for a compiler-assisted and enforced check, it would be perhaps less globally conflated to set a member property per the feature of this issue versus setting up the global `prototype` chain of a constructor function. Yet wouldn't the TypeScript code which depends on this `instanceof` be forced to import the module for the declaration of the type it is employing, thus wouldn't the `prototype` chain be guaranteed to exist in the global scope? I mean if the caller can violate the expected type, then he can violate `List` and `BaseCollection` also, not just the interfaces which TypeScript doesn't currently add to the `prototype` chain.

So again I arrive at the same conclusion, which is that if you want to use TypeScript typing, then you have no guarantees about runtime soundness any way, unless you can assume TypeScript has been employed globally in your ECMAScript universe. So I see no benefit destroying the soundness of the internal consistency of the TypeScript type system. If the caller isn't interopting with the TypeScript type system where the callee is depending on it, then there will be runtime errors. Enabling some heuristic convenience user-guard so that callers which don't want to set the `prototype` chain can instead set some member property on the specific instance, is moving us towads utter chaos if such heuristics proliterate. Where do you draw a line and stop creating special case interactions which render the type system inconsistent?

**Comment by shelby3** [comment]
I offered an idea about populating the `prototype` chain with interfaces as a more JavaScript idiomatic paradigm (seems that is the purpose of `prototype, thus the idiomatic place to tag the interface structure) and which doesn't break the internal consistency of the type system, as an alternative to this issue's adhoc choice of setting member properties on specific instances without any enforcement by JavaScript construction mechanism. For me, that already appears to be a slamdunk.

There is another way to do this though. That is for the compiler to make a new type check (not `instanceof`) which has the semantics of checking what the compiler knows which may be erased at runtime. And use this new feature (keyword?) to do guards. The JavaScript runtime would be completely oblivious to this erased information and the non-TypeScript code wouldn't be able perform these checks at runtime.

**Comment by aluanhaddad** [comment]
> Please do not chase away dissenting technical opinions. You need them. Consider the value of learning to be tolerant of others and value their technical disagreement. Greatness requires it.

I couldn't agree more. However, it's perfectly valid to program with functions and object literals and to never use classes or even manually wired inheritance via prototypes.

> Simply match on structure at the guard instead (and use the result of the match to narrow the type). So you'd need some new compiler provided function that performs this structural match similar to my second paragraph in this comment. And at compile-time, this function need not be called because the compiler can infer the type. At run-time, the function would actually need to be called to do a structural matching to check for the interface type .

What specifically do you have in mind? Type cards are most often used when type information has been lost because it comes from untyped APIs and unknown sources. If the compiler could always determine the actual type, we wouldn't be having this discussion but that is simply not possible.

> And do note that even on the untyped (uni-typed) ECMAScript, the prototype chain is form of heritage of objects since it is global to all constructed instances which didn't override the prototype as constructed (and even retroactively so which is the point I want to explore modeling with typeclasses for potentially amazing benefits in productivity and extensibility).

I would definitely be curious to see your formal type class proposal. 
I think it is a very interesting point that you bring up here about retroactive modification of the prototype chain. Since the prototype property of an object and the prototype's properties are in fact frequently mutable, an object can change at any point during runtime such that it no longer matches it's declared static shape. Furthermore, irrespective of mutability, you can create an object which will pass the check `o instanceof Date`, but not conform to Date's behavior. This is because JavaScript's `instanceof` operator has no notion of types it only understands values. Walking the prototype chain and doing reference comparisons is far from reliable.

I think that as you have yourself stated, a different language, one targeting a type aware VM may well be the only way to achieve your goals.

That said, you really should spend some more time with the language and see if it is in fact useful to you before continuing.
One important thing to be aware of is that
TypeScript's classes are structurally typed not normally typed. This is a structurally typed language.

**Comment by spion** [comment]
@shelby3 

First, you can't populate the prototype chain with interfaces. This will not work: specifically it  `instanceof` does not work reliably cross-realm. By that I mean e.g. `a instanceof Array` can return false if the array `a` came from e.g.  another iframe (more generally, another realm). See [this discussion](https://groups.google.com/forum/#!msg/comp.lang.javascript/XTWYCOwC96I/70rNoQ3L-xoJ)

Secondly, npm (the most popular package manager for JavaScript at the moment) compounds the above problem by installing multiple semver-incompatible versions of the same library into different directories. This in turn means that a class defined in such a module will actually have more than one value; and again instanceof won't work reliably.

Finally, this is simply not how most JavaScript is written. Infact it cannot be how most JS is written, as this would be a typescript-only feature. And here we come to a clash to TypeScript's design goals: to be a type system which helps with existing JavaScript code. Just look at how [Promises/A+](https://promisesaplus.com/) thenables are specified. Its all about a method `then` present on the thenable object. Not about some non-existing constant "Thenable" that should be in the prototype chain.

As to why type guards are okay, I'll just quote myself without the "you are wrong" part:

> This sort of feature is precisely what makes TypeScript different from other typed languages. While they decide NOT to trust the user unless their input fits their narrow preconceptions of what is correctly modelled code, TypeScript takes a different approach. TypeScript mostly trusts the user. The user, in turn, needs to take special care in ensuring some of their code is correct, like type guards. But not all, as it would be with dynamic languages.
>
> This is the biggest win of TypeScript: it lets you pick between "great power, great responsibility" (carefully hand-check the code, ensure that your type assertions are correct) and "lesser power, lesser responsibility" (once you hand-"prove" type assertions are correct, you don't have to do it for the rest of the code).
>
> Its not a type system that ensures correctness. Its a type system that works with the user to ensure it. Its a "bring your own lemma" type system

So this feature is sound in this sense: "I can't automatically check this, but if you supply your own proof that the type is indeed correct, I will accept that"

**Comment by shelby3** [comment]
@aluanhaddad wrote:
> However, it's perfectly valid to program with functions and object literals and to never use classes or even manually wired inheritance via prototypes.

Agreed. That is why I mentioned the purely structural option instead of the feature of this issue which was adopted.

It is difficult to have an open discussion and ignore discussion. Therefor...

<sup>First, I am unblocking you (perhaps contrary to my better judgement) because it seems I can trust you to talk technicals (and you may have valuable discussion to share) and to not to involve me in discrimination claims. If that changes, I may regrettably be forced to backtrack. I am not stating this as if I desire any authority or control over you (nor to insinuate any judgement of blame or correctness), rather this is just a statement of my personal policy w.r.t. to you. Please avoid making any insinuations that would cause me to consider a legal protection stance to be more important than open discussion. For me, open discussion is paramount, but I do have to be pragmatic in this age where we all commit 7 felonies per day just by breathing.</sup>

>> Simply match on structure at the guard instead (and use the result of the match to narrow the type). So you'd need some new compiler provided function that performs this structural match similar to my second paragraph in this comment. And at compile-time, this function need not be called because the compiler can infer the type. At run-time, the function would actually need to be called to do a structural matching to check for the interface type .

> What specifically do you have in mind? Type guards are most often used when type information has been lost because it comes from untyped APIs and unknown sources. If the compiler could always determine the actual type, we wouldn't be having this discussion but that is simply not possible.

```ts
if (someNode.isA(Sortable) {
    someNode.sort()
}
```

The compiler would emit:

```ts
if (someNode.isA({ sort:function() {} }) {
    someNode.sort()
}
```

So the `isA` function would check that the properties of the `interface` match structurally up the capabilities of what the runtime can check, e.g. the existence of the `sort` property, that it has a `typeof == 'function'`, and the number of parameters of the function.

That seems to be much more sane than the feature that was adopted, because at least it enforces structural type checking at compile-time and even marginal structural type checking at runtime.

Note if there is no `else` case on the guard, I presume by default it should throw an exception at runtime if the `if` condition is `false`.

**Comment by shelby3** [comment]
The following concerns the nominal typing idea I promulgated in this thread, which is orthogonal to the prior comment of mine explaining a purely structural idea.

Note I have to rush out to dinner, and this comment will be refined and augmented when I return. Just very quickly...

@spoin wrote:
> First, you can't populate the prototype chain with interfaces. This will not work because instanceof does not work reliably cross-realm.

Structural can fail also due to false positive matches (both at compile-time and runtime). Nominal typing can both fail dynamically at runtime, but not at compile-time.

Choose your poison.

> The rest of the reasons are complex, but mainly its a combination of "no single module system"

Prototype inheritance is inherently modular.

**Comment by spion** [comment]
> I do not consider these potential pitfalls with nominal typing to be a rational justification to completely avoid nominal typing with JavaScript. As I wrote above, structural typing also has pitfalls. Programmers should have both nominal and structural typing in their toolchest.

You can consider it whatever you want, the fact is that the kind of JS that developers normally write is based on basic structural checking, and as such its TypeScript's primary job to support that.

Additionally, nominal type systems suffer from the "interface afterthought" problem. Example:

1. First promise library P1 appears, written by author A1
2. A1 never gave second thought to the "Thenable" idea, so they never wrote a separate "Thenable" (interface) module. 
3. Another author A2 writes a better promise library P2. Due to the nature of nominal types, its incompatible with A1
4. Because of incompatibility with P1, P2 cannot achieve enough traction and adaptation without explicitly depending on P1 or inheriting from P1, as a lot of things use P1, and P1 uses nominal checking to ensure correct types.
4. Frustrated by this, A2 writes Thenable module, and urges A1 to adopt it.
5. A1 sees no value in this, due to the fact that library P2 doesn't have enough traction to justify a change
6. The development of promises has stalled.

Regarding your structural check idea, can you please tell me what the cost will be to check the following?

```ts
interface Node {
  __special_tag_to_check_if_value_is_node: string;
  data: <T>
  children: Array<Node<T>>
}
```

**Comment by shelby3** [comment]
@spion wrote:
> You can consider it whatever you want, the fact is that the kind of JS that developers normally write is based on basic structural checking, and as such its TypeScript's primary job to support that.

Please re-read my prior comment as I have rebutted this "normally" argument.

> Additionally, nominal type systems suffer from the "interface afterthought" problem. Example:

You are conflating nominal typing with subclassing. That is why I am preparing to promulgate typeclasses type of the `prototype` chain. I am about to radically impact the JavaScript universe on a large scale. TypeScript can come along for the ride or it can be obstinate. Either way, I am going to see this concept gets implemented (eventually). I've been working on this concept for past several years (on and off) and very intensely this past May.

**Comment by spion** [comment]
> I do not consider these potential pitfalls with nominal typing to be a rational justification to completely avoid nominal typing with JavaScript. As I wrote above, structural typing also has pitfalls. Programmers should have both nominal and structural typing in their toolchest.

The difference in pitfalls is fundamental. With `instanceof` checks, code that is *supposed to work* breaks. With structural checks, code that is not supposed to work breaks at run time, rather than compile time.

**Comment by shelby3** [comment]
@spion wrote:
> The difference in pitfalls is fundamental. With `instanceof` checks, code that is _supposed to work_ breaks. With structural checks, code that is not supposed to work breaks at run time, rather than compile time.

That is an interesting perspective, but it depends on who and what was "supposed to". If `instanceof` breaks, is it because the programmer was supposed to be aware of the couple of general ways it can fail and avoid them? So then was it supposed to work or not supposed to work?

Your logic presumes that structural code fails because the programming designed the code wrong, but wasn't he supposed to design it correctly? And structural code can fail at compile-time, if we presume that not having the ability to distinguish between intent and structure as a failure of compile-time typing.

I hope you are somewhat convinced that your choices were arbitrary.

**Comment by spion** [comment]
> Prototype inheritance is inherently locally coherent and modular because it is object-based, so there doesn't need to be any global coherence. Could you please explain more clearly what problem you envision?

Promises/A+ thenables are a good example. If you want to specify a `Thenable` nominal interface, there needs to be a single value that represents it in the prototype chain (in order for `instanceof` to work). To get this single value into all libraries that implement `Thenable`, it needs to be a module of a module system that guarantees a single instance will be delivered when requested via import. AFAIC This is not guaranteed by either ES6 modules or CommonJS modules, so at best you would need to ensure it in the module loader spec, and any environment that uses different loaders as well (nodejs?)

Btw, ES6 tried and failed to solve this problem (for users) with Symbols. The final solution ended up being a string-addressed [global symbol registry](http://www.2ality.com/2014/12/es6-symbols.html#crossing_realms_with_symbols).

**Comment by spion** [comment]
@shelby3 

> That is an interesting perspective, but it depends on who and what was "supposed to". If instanceof breaks, is it because the programmer was supposed to be aware of the couple of general ways it can fail and avoid them? So then was it supposed to work or not supposed to work?

That would mean avoiding `instanceof` to check arguments passed externally, as they may come from another realm. Which means avoiding its use as a type guard.

**Comment by shelby3** [comment]
@spion wrote:
>> That is an interesting perspective, but it depends on who and what was "supposed to". If `instanceof` breaks, is it because the programmer was supposed to be aware of the couple of general ways it can fail and avoid them? So then was it supposed to work or not supposed to work?

> That would mean avoiding `instanceof` to check arguments passed externally, as they may come from another realm. Which means avoiding its use as a type guard.

Or avoiding the other realms.

**Comment by RyanCavanaugh** [comment]
Previous comments here have been running afoul of the [Code of Conduct](https://opensource.microsoft.com/codeofconduct/), but I appreciate everyone redirecting their attention to the technical discussion at hand. Let's keep it that way.

**Comment by shelby3** [comment]
@spion wrote:
>> I hope you are somewhat convinced that your choices were somewhat arbitrary.
>
> I'm convinced you are willing to employ sophistry to push your argument.

Another false accusation without a quote. What fallacious argument have I employed?

This is about 4 or 5 times you've resorted to frustration when you can't get me to admit that you are entirely correct and I am entirely wrong.

**Comment by shelby3** [comment]
@spion wrote:
>> Prototype inheritance is inherently locally coherent and modular because it is object-based, so there doesn't need to be any global coherence. Could you please explain more clearly what problem you envision?

> Promises/A+ thenables are a good example. If you want to specify a `Thenable` nominal interface, there needs to be a single value that represents it in the prototype chain (in order for `instanceof` to work). 

If we are referring to subclassing and not typeclasses, the `Promise` constructor function controls what will be put in the `prototype` chain. Even if you import multiple instances of a `Promise`, they will all for each use only one `Thenable` interface per `prototype` chain. But with redundant imports, all of these `Thenable` interfaces will not have the same instance pointer, since the `Promise` constructor function would have multiple instances. So I agree that non-redundant imports are necessary if we expect to have a unified `instanceof`for all instances.

> To get this single value into all libraries that implement `Thenable`, it needs to be a module of a module system that guarantees a single instance will be delivered when requested via import.

Yes and I designed such an import system for my coding, but it doesn't solve the cross-realm issue. And this would require me to be sure all libraries I use which can pass my code a `Promise` also use my importing system.

A consistent importing system wide is important. I agree.

>  AFAIC This is not guaranteed by either ES6 modules or CommonJS modules, so at best you would need to ensure it in the module loader spec, and any environment that uses different loaders as well (nodejs?)

I understand there is a lot of broken legacy realms. C'est la vie. We move forward anyway.

> Btw, ES6 tried and failed to solve this problem (for users) with Symbols. The final solution ended up being a string-addressed [global symbol registry](http://www.2ality.com/2014/12/es6-symbols.html#crossing_realms_with_symbols).

Instead of string keys, they could have used 126-bit cryptographic hashes (or any approximation to a random oracle) to be [probabilistically sure of no collisions](http://preshing.com/20110504/hash-collision-probabilities/)

Did it fail because of name space collisions, lack of adoption, or what?

**Comment by spion** [comment]
> Or avoiding the other realms. We shouldn't presume the only use of ECMAScript is in broken realms such as the browser and NPM. The (expanse of possibilities in the unpredictable future of the) universe is not so tiny.

This is not what realms means. A realm can be thought of as a fresh global "scope". For example an iframe has a different "realm" from its parent window. That means it has different unique global (window) object, as well as other globals: e.g. Array constructor and array prototype. As a result, passing an array you got from an iframe to a function that checks `instanceof Array` means the array will fail the check.

In CommonJS, every module is wrapped by a header and footer that form a closure:

```js
function(module, exports, ...) {

<module code here>

}
```

Which is then called with an empty exports and initialized module object, at discretion, by the module system.

Its the same problem with e.g. class expressions:

```js
function makeClass() {
  return class C {
    // definition here
  }
}

let C1 = makeClass(), C2 = makeClass(), c1 = new C1(), c2 = new C2();

assert(c1 instanceof C2) // fails
assert(c2 instanceof C1) // fails

> Did it fail because of name space collisions, lack of adoption, or what?

It failed because they came up with a neat little way to define unique constants that don't clash with anything existing and can be used as object keys, then wanted to expose this mechanism to users somehow, but failed to take cross-realm issues into account. So we're back to string keys, which is where we were in the first place before Symbols entered the scene.

By all means, take 160-bit cryptographic hashes idea to esdiscuss. May I ask though, exactly what is the thing that you plan to hash to get the unique key that solves the multi-realm problem?

**Comment by yortus** [comment]
As of ES6, `instanceof` is decoupled from the prototype chain due to [`Symbol.hasInstance`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/hasInstance). Walking the prototype chain is now just the *default* behaviour. But in `obj instanceof Obj`, if the `Obj` value has its own `[Symbol.hasInstance]` property, then that will determine how `instanceof` behaves.

**Comment by eggers** [comment]
@shelby3 TypeScript wasn't meant to be a new language. It was meant to just add types on top of Javascript. One of the current trends of JavaScript is something called DuckTyping. (If it walks like a Duck and quacks like a Duck, then it's a Duck.) That is a very different concept than inheritance and is in fact antithetical to it. Interfaces are TypeScripts answer to DuckTyping. Putting Interfaces on the prototype chain would defeat their purpose. User defined type guards were meant to solve type guards for interfaces. Maybe there is a better way of creating them. (I personally would prefer that they were more tightly bound to the interfaces themselves.) However, user defined type guards definitely belong in TypeScript, and they definitely don't belong on the prototype chain.



```ts
/* DuckTyping */

interface Foo { foo:string };

function bar(foo: Foo) {
  // something
}

var foo = {foo: 'bar'};
bar(foo); // Legal even though Foo was never explicitly implemented.
```


```ts
/* Multiple Inheritance */

interface Car {
  goto(dest: Land): void;
}

interface Boat {
  goto(des: Water): void;
}

class HoverCraft {
  goto(dest: Water|Land) {
    // something
  }
}
```

@yortus That's awesome. Maybe a separate mechanism for user defined type guards could be implemented that could be down compiled as the current type guards are. I personally think that it would be more intuitive to do write something like this (The type guards should be more closely bound to the interfaces than they currently are):

```ts
interface Cat {
  name: string;
  static [Symbol.hasInstance](animal: Animal) {
    return a.name === 'kitty';
  }
}

if (c instanceof Cat) { // or maybe `c implements Cat`
  // dog barks.
}
```

which could could compile to:

```es6
// es6
class Cat {  
  static [Symbol.hasInstance](instance) {
    return a.name === 'kitty';
  }
}

if (c instanceof Cat) {
  // dog barks.
}

// es5

var Cat = (function () {
    function Cat() {
    }
    Cat[Symbol.hasInstance] = function (instance) {
        return a.name === 'kitty';
    };
    return Cat;
}());

if (Cat[Symbol.hasInstance(c)) {
  // dog barks.
}
```

**Comment by shelby3** [comment]
@spion wrote:
>> Or avoiding the other realms. We shouldn't presume the only use of ECMAScript is in broken realms such as the browser and NPM. The (expanse of possibilities in the unpredictable future of the) universe is not so tiny.
>
> This is not what realms means.

I did not define 'realms'.

> A realm can be thought of as a fresh global "scope". For example an iframe has a different "realm" from its parent window.

It is evident that I knew that by noticing that "avoiding" those could involve "avoiding ... broken realms such as [in] the browser".

**Comment by shelby3** [comment]
@eggers please note I have made three different possible suggestions to choose from. One of them is to [use purely structural matching for the user guard](https://github.com/Microsoft/TypeScript/issues/1007#issuecomment-246318431) (which appears to fix the serious soundness flaws that this issue's "fix" created), so obviously I am not advocating putting any interface in the `prototype` chain for that suggestion.

**Comment by shelby3** [comment]
@aluanhaddad wrote:
> I would definitely be curious to see your formal type class proposal. 
> ...
> Walking the prototype chain and doing reference comparisons is far from reliable.
>
> I think that as you have yourself stated, a different language, one targeting a type aware VM may well be the only way to achieve your goals.

To the extent that TypeScript can embrace unreliability of expected types at runtime (even possibly with optional runtime nominal and structural checks that defer to a default or error case), I think the similar level of typing assurances can be attained with typeclasses.

And I believe typeclasses are much more flexible for extension, compared to subclassing. I intend to attempt to explain that soon in issue thread I created for that.

**Comment by eggers** [comment]
@shelby3 Ah, I missed that structural proposal. There has been a lot to read the last couple of days in here.

I actually do think something like that would work. It would add some runtime overhead for a large interface as checking for the existence of many fields would take some time, but probably not prohibitive. By default the TypeScript compiler could out put code that checked for all properties/functions, with the option of overriding `[Symbol.hasInstance]` with a custom check. However rather than using a special function on interfaces `isA(x)`, I would use a keyword like `implements` or `implementationOf`, maybe overloading `instanceOf`.

**Comment by shelby3** [comment]
@eggers wrote:
> By default the TypeScript compiler could out put code that checked for all properties/functions, with the option of overriding `[Symbol.hasInstance]` with a custom check.

Also, when the compiler constructed the instance within the function, then it could optimize away the runtime structural check.

**Comment by eggers** [comment]
@shelby3 can you explain more about how it could optimize away the structural check? If you need to do one thing if it's an implementation of `Animal` and another if it's one of `Vehicle`, you still need to structurally check which object it implements.

**Comment by shelby3** [comment]
@eggers when the compiler knows that the instance was constructed within the function, then it knows at runtime it has to be of the type that was constructed, thus it doesn't need to do any runtime check for the structural type.

Also I want to add that ideas for tagging the structural type instead of checking its structure, such as the feature that was implemented for this issue #1007, are breaking structural type checking. And the feature of #1007, also breaks the internal consistency of the compiler.

Today I am going to be initiating+participating in a more holistic analysis of all this and tying it into the nominal typing discussion, as well as my proposal for typeclasses. I'll try to remember to cross-link from this issue discussion.

**Comment by nmaro** [comment]
Is there a way to define a type guard that activates a type *if it returns*? Something that would allow you to write something like this:

```
try {
   checkIsA(o)
   // from here on o has type A
} catch(e) {
}
```

**Comment by kitsonk** [comment]
This is not a support forum.

Questions should be asked at [StackOverflow](http://stackoverflow.com/questions/ask?tags=typescript) or on [Gitter.im](https://gitter.im/Microsoft/TypeScript).

**Comment by yortus** [comment]
@nmaro it has been suggested but not implemented so far. See for example #8655, and other issues linked from there.
